mpitests/test_apply.py

```diff
@@ -439,32 +485,527 @@ def test_dev_reset(self, tol, operation, par, Wires):
-        def circuit_mpi():
-            qml.PauliX(wires=[0])
-            qml.PauliX(wires=[0])
-            return qml.state()
-
+        gpumpi_qnode = qml.QNode(circuit, dev_gpumpi)
         dev_gpumpi.reset()
 
-        local_state_vector = circuit_mpi()
+        local_state_vector = gpumpi_qnode()
         assert np.allclose(local_state_vector, local_expected_output_cpu, atol=tol, rtol=0)
+
+
+class TestExpval:
+    """Tests that expectation values are properly calculated or that the proper errors are raised."""
+
+    @pytest.mark.parametrize(
+        "operation",
+        [
+            qml.PauliX,
+            qml.PauliY,
+            qml.PauliZ,
+            qml.Hadamard,
+            qml.Identity,
+        ],
+    )
+    @pytest.mark.parametrize("wires", [0, 1, 2, numQubits - 3, numQubits - 2, numQubits - 1])
+    def test_expval_single_wire_no_parameters(self, tol, operation, wires):
+        """Tests that expectation values are properly calculated for single-wire observables without parameters."""
+
+        obs = operation(wires)
+        expval_single_wire_no_param(tol, obs)
+
+    @pytest.mark.parametrize(
+        "obs",
+        [
+            qml.PauliX(0) @ qml.PauliZ(1),
+            qml.PauliX(0) @ qml.PauliZ(numQubits - 1),
+            qml.PauliX(numQubits - 2) @ qml.PauliZ(numQubits - 1),
+            qml.PauliZ(0) @ qml.PauliZ(1),
+            qml.PauliZ(0) @ qml.PauliZ(numQubits - 1),
+            qml.PauliZ(numQubits - 2) @ qml.PauliZ(numQubits - 1),
+        ],
+    )
+    def test_expval_multiple_obs(self, obs, tol):
+        """Test expval with Hamiltonian"""
+        num_wires = numQubits
+        comm = MPI.COMM_WORLD
+
+        dev_cpu = qml.device("default.qubit", wires=num_wires, c_dtype=np.complex128)
+        dev_gpumpi = qml.device("lightning.gpu", wires=num_wires, mpi=True, c_dtype=np.complex128)
+
+        def circuit():
+            qml.RX(0.4, wires=[0])
+            qml.RY(-0.2, wires=[numQubits - 1])
+            return qml.expval(obs)
+
+        cpu_qnode = qml.QNode(circuit, dev_cpu)
+        gpumpi_qnode = qml.QNode(circuit, dev_gpumpi)
+
+        assert np.allclose(cpu_qnode(), gpumpi_qnode(), atol=tol, rtol=0)
+
+    @pytest.mark.parametrize(
+        "obs, coeffs",
+        [
+            ([qml.PauliX(0) @ qml.PauliZ(1)], [1.0]),
+            ([qml.PauliX(0) @ qml.PauliZ(numQubits - 1)], [1.0]),
+            ([qml.PauliZ(0) @ qml.PauliZ(1)], [1.0]),
+            ([qml.PauliZ(0) @ qml.PauliZ(numQubits - 1)], [1.0]),
+            ([qml.PauliX(0) @ qml.PauliZ(1), qml.PauliZ(0) @ qml.PauliZ(1)], [1.0, 0.2]),
+            (
+                [qml.PauliX(0) @ qml.PauliZ(numQubits - 1), qml.PauliZ(0) @ qml.PauliZ(1)],
+                [1.0, 0.2],
+            ),
+            (
+                [
+                    qml.PauliX(numQubits - 2) @ qml.PauliZ(numQubits - 1),
+                    qml.PauliZ(0) @ qml.PauliZ(1),
+                ],
+                [1.0, 0.2],
+            ),
+        ],
```

Comment:
I think it would be best to use coeffs with values other than 1.0 here to ensure testing doesn't hit magic number combinations.

Reply:
Sure!